core/src/test/kotlin/com/kin/ecosystem/core/data/order/ExternalEarnOrderCallTest.kt

```diff
@@ -0,0 +1,107 @@
+import com.kin.ecosystem.core.network.model.OpenOrder
+import com.kin.ecosystem.core.network.model.Order
+import com.nhaarman.mockitokotlin2.*
+import kin.ecosystem.test.base.BaseTestClass
+import kin.sdk.migration.common.KinSdkVersion
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.annotation.Config
+import java.math.BigDecimal
+
+@Config(manifest = Config.NONE)
+@RunWith(RobolectricTestRunner::class)
+class ExternalEarnOrderCallTest : BaseTestClass() {
+
+	private val orderDataSource: OrderDataSource = mock {
+		on { createExternalOrderSync(any()) } doAnswer { openOrder}
+	}
+	private val blockchainSource: BlockchainSource = mock {
+		on { blockchainVersion } doAnswer { KinSdkVersion.OLD_KIN_SDK }
+	}
+	private val eventLogger: EventLogger = mock()
+	private val externalOrderCallbacks: CreateExternalOrderCall.ExternalOrderCallbacks = mock()
+
+	private val openOrder: OpenOrder = mock {
+		on { offerId } doAnswer  { offerId }
+		on { id } doAnswer { orderId }
+		on { amount } doAnswer { 10 }
+		on { blockchainData } doAnswer { blockchainData }
+	}
+
+	private val blockchainData: BlockchainData = mock {
+		on { recipientAddress } doAnswer { "some_address" }
+	}
+
+	private val order: Order = mock {
+		on { orderId } doAnswer { orderId }
+		on { offerId } doAnswer { offerId }
+		on { result } doAnswer { jwtBodyPaymentConfirmationResult }
+	}
+
+	private val payment: Payment = mock {
+		on { orderID } doAnswer { orderId }
+		on { isSucceed } doAnswer { true }
+		on { transactionID } doAnswer { transactionId }
+		on { type } doAnswer { Payment.EARN }
+	}
+
+	private val jwtBodyPaymentConfirmationResult: JWTBodyPaymentConfirmationResult = mock {
+		on { jwt } doAnswer { "jwt_confirmation" }
+	}
+
+
+	private lateinit var externalEarnOrderCall: ExternalEarnOrderCall
+
+	@Before
+	override fun setUp() {
+		super.setUp()
+		externalEarnOrderCall = ExternalEarnOrderCall(orderDataSource, blockchainSource, orderJwt, eventLogger, externalOrderCallbacks)
+	}
+
+	@Test
+	fun `run kin2 happy flow to completion`() {
+		externalEarnOrderCall.run()
+		val paymentCaptor = argumentCaptor<Observer<Payment>>()
+		inOrder(orderDataSource, blockchainSource, eventLogger, externalOrderCallbacks).apply {
+			verify(orderDataSource).createExternalOrderSync(orderJwt)
+			verify(eventLogger).send(any<EarnOrderCreationReceived>())
+			verify(blockchainSource).addPaymentObservable(paymentCaptor.capture())
+			verify(blockchainSource).blockchainVersion
+			argumentCaptor<KinCallback<Order>>().apply {
+				verify(orderDataSource).submitEarnOrder(any(), isNull(), any(), capture())
+				firstValue.onResponse(order)
+			}
+			verify(eventLogger).send(any<EarnOrderCompletionSubmitted>())
+			paymentCaptor.firstValue.onChanged(payment)
+			verify(blockchainSource).removePaymentObserver(any())
+			argumentCaptor<KinCallback<Order>>().apply {
```

Comment:
can you refactor everything to be sync? I think it will be more readable and testable this way

Reply:
yes, I will but not on this task.